### PR TITLE
2269 publish events from log4shell exploiter

### PR DESCRIPTION
**Pull request by cakekoa**
# What does this PR do?

Fixes part of #2269.

Update log4shell exploiter to publish propagation and exploitation events.

## PR Checklist
* [x] Have you added an explanation of what your changes do and why you'd like to include them?
* [x] Is the TravisCI build passing?
* [x] ~~Was the CHANGELOG.md updated to reflect the changes?~~
* [x] ~~Was the documentation framework updated to reflect the changes?~~
* [x] Have you checked that you haven't introduced any duplicate code?

## Testing Checklist

* [x] ~~Added relevant unit tests?~~
* [ ] ~~Have you successfully tested your changes locally? Elaborate:~~
    > Tested by {Running the Monkey locally with relevant config/running Island/...}
* [ ] ~~If applicable, add screenshots or log transcripts of the feature working~~

### CODE REVIEW

File: monkey/infection_monkey/exploit/log4shell.py [hunk 1]
```diff
@@ -147,10 +157,17 @@ def exploit(self, url, command) -> None:
                 )
                 try:
                     url = exploit.trigger_exploit(self._build_ldap_payload(), self.host, port)
-                except Exception as ex:
-                    logger.warning(
-                        "An error occurred while attempting to exploit log4shell on a "
-                        f"potential {exploit.service_name} service: {ex}"
+                except Exception as err:
+                    error_message = "An error occurred while attempting to exploit log4shell on a "
+                    f"potential {exploit.service_name} service: {err}"
```
Comment (mssalvatore) [inline monkey/infection_monkey/exploit/log4shell.py:162]: This isn't working as expected. Run the below example as a test:

```python
import time

error_message = "The first part of the "
f"error message {time.time()}"

print(error_message)
```

Fix with:
 ```suggestion
                    error_message = (
                        "An error occurred while attempting to exploit log4shell on a "
                        f"potential {exploit.service_name} service: {err}"
                    )
```

Black will probably have something to say about that formatting.

File: monkey/infection_monkey/exploit/log4shell.py [hunk 2]
```diff
@@ -147,10 +157,17 @@ def exploit(self, url, command) -> None:
                 )
                 try:
                     url = exploit.trigger_exploit(self._build_ldap_payload(), self.host, port)
```
Comment (mssalvatore) [inline monkey/infection_monkey/exploit/log4shell.py:159]: I think timestamp for exploitation should be recorded here.

File: monkey/infection_monkey/exploit/log4shell.py [hunk 3]
```diff
@@ -187,6 +216,11 @@ def _wait_for_victim_to_download_agent(self):
 
         while not timer.is_expired():
             if self._agent_http_server_thread.downloads > 0:
+                self._publish_propagation_event(
+                    target=self.host.ip_addr,
+                    propagation_success=True,
+                    tags=PROPAGATION_TAGS,
+                )
```
Comment (mssalvatore) [inline monkey/infection_monkey/exploit/log4shell.py:223]: Using the current timestamp actually makes sense here, I think.

👍

**Review by mssalvatore — CHANGES_REQUESTED** [review]
I'm torn on whether the event publishing should go in `_wait_for_victim()` or stay where it is down in the different wait methods. Let's take another look once we shorten the publish calls.

File: monkey/infection_monkey/exploit/log4shell.py [hunk 4]
```diff
@@ -115,7 +123,9 @@ def _build_ldap_payload(self) -> str:
 
     def _build_command(self, path: PurePath, http_path) -> str:
         # Build command to execute
-        monkey_cmd = build_monkey_commandline(self.servers, self.current_depth + 1, location=path)
+        monkey_cmd = build_monkey_commandline(
+            self.servers, self.current_depth + 1, location=str(path)
+        )
```
[the remark this answers is not in the document]
Comment (cakekoa) [reply monkey/infection_monkey/exploit/log4shell.py:128]: It _would_ if `build_monkey_commandline()` wasn't lying about its interface -- it actually accepts any argument and calls `str()` on it

**Review by shreyamalviya — COMMENTED** [review]
Same comments as https://github.com/guardicore/monkey/pull/2401

**Pull request merged**